From af245d08ad8571344b4f72025541281586010484 Mon Sep 17 00:00:00 2001
From: Hong-Mei Li <a21834@motorola.com>
Date: Thu, 24 Sep 2015 18:21:24 -0700
Subject: [PATCH 189/959] IKSWM-4458 power:reset: Force system cold/warm reset

1. Force cold reboot after command "adb reboot bootloader"
2. Force warm reboot after kernel panic

Merged commit:
a7cc0f7e6e138884e0b84106858bf130a3084f85
a55e84f5d0b682fb8d49ecc7c25762470f23a163

Change-Id: Iaa49a7eaecaefa4ef43d3ed1f9ad5ebf5c7e6a59
Signed-off-by: Stanley Lei <a23075@motorola.com>
Reviewed-on: http://gerrit.pcs.mot.com/625574
SLTApproved: Slta Waiver <sltawvr@motorola.com>
Tested-by: Jira Key <jirakey@motorola.com>
Reviewed-by: Qing Chang <qing@motorola.com>
Reviewed-by: Yi-Wei Zhao <gbjc64@motorola.com>
Submit-Approved: Jira Key <jirakey@motorola.com>
Signed-off-by: Hong-Mei Li <a21834@motorola.com>
---
 drivers/power/reset/msm-poweroff.c | 10 +++++++++-
 include/linux/qpnp/power-on.h      |  1 +
 2 files changed, 10 insertions(+), 1 deletion(-)

diff --git a/drivers/power/reset/msm-poweroff.c b/drivers/power/reset/msm-poweroff.c
index cbc979e1944..447a05f80e3 100644
--- a/drivers/power/reset/msm-poweroff.c
+++ b/drivers/power/reset/msm-poweroff.c
@@ -246,7 +246,7 @@ static void msm_restart_prepare(const char *cmd)
 	}
 
 	/* Hard reset the PMIC unless memory contents must be maintained. */
-	if (need_warm_reset) {
+	if (need_warm_reset || in_panic) {
 		qpnp_pon_system_pwr_off(PON_POWER_OFF_WARM_RESET);
 	} else {
 		qpnp_pon_system_pwr_off(PON_POWER_OFF_HARD_RESET);
@@ -257,6 +257,14 @@ static void msm_restart_prepare(const char *cmd)
 			qpnp_pon_set_restart_reason(
 				PON_RESTART_REASON_BOOTLOADER);
 			__raw_writel(0x77665500, restart_reason);
+			/* set reboot_bl flag in PMIC for cold reset */
+			qpnp_pon_store_extra_reset_info(RESET_EXTRA_REBOOT_BL_REASON,
+				RESET_EXTRA_REBOOT_BL_REASON);
+			/*
+			 * force cold reboot here to avoid impaction from
+			 * modem double reboot workaround solution.
+			 */
+			qpnp_pon_system_pwr_off(PON_POWER_OFF_HARD_RESET);
 		} else if (!strncmp(cmd, "recovery", 8)) {
 			qpnp_pon_set_restart_reason(
 				PON_RESTART_REASON_RECOVERY);
diff --git a/include/linux/qpnp/power-on.h b/include/linux/qpnp/power-on.h
index 11a088d9228..5c6c38d835c 100644
--- a/include/linux/qpnp/power-on.h
+++ b/include/linux/qpnp/power-on.h
@@ -55,6 +55,7 @@ enum pon_restart_reason {
 };
 
 #define RESET_EXTRA_PANIC_REASON	BIT(3)
+#define RESET_EXTRA_REBOOT_BL_REASON	BIT(2)
 
 #ifdef CONFIG_QPNP_POWER_ON
 int qpnp_pon_system_pwr_off(enum pon_power_off_type type);
-- 
2.11.0

